Issue 6764

JustinKyleJames opened the pull request:
This is to fix #6764.

alanking commented:
I'm going to adjust the PR message so that it doesn't close the issue when we merge this (see https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword)

alanking commented on plugins/database/src/db_plugin.cpp:
```diff
@@ -7839,7 +7839,8 @@ irods::error db_mod_user_op(
         char userIdStr[MAX_NAME_LEN];
         i = decodePw( _ctx.comm(), _new_value, decoded );
-        if (strlen(decoded) > MAX_PASSWORD_LEN - 8) {
+        if (i == CAT_PASSWORD_ENCODING_ERROR || strlen(decoded) > MAX_PASSWORD_LEN - 8) {
```
Is there ever a case where `CAT_PASSWORD_ENCODING_ERROR` is returned but it's not because the password is too long? What if the password is too short (<3 characters)?

JustinKyleJames replied:
We discussed this in standup the other day.  It is hard to say due to all the stuff that is going on in decodePw.  That possibility is why we decided to not update decodePw but instead catch the error here.

alanking replied:
@JustinKyleJames - I still think a comment would be helpful here, but if we're not going to do that, please resolve this conversation. Thanks!

korydraughn replied:
Agreed.

@JustinKyleJames If you've acquired knowledge about this, let's write it down for the next person. This just helps us move faster.

JustinKyleJames replied:
I added a comment for this.  I said that the CAT_PASSWORD_ENCODING_ERROR occurs when the password is not of the correct length.  There may be other cases where a CAT_PASSWORD_ENCODING_ERROR isn't due to that but I am not able to determine from the code if that is the case.

korydraughn commented on scripts/irods/test/test_iadmin.py:
```diff
@@ -2226,6 +2226,28 @@ def test_moduser_password(self):
+        # Change password to 45 bytes (should fail)
+        pw4 = '123456789012345678901234567890123456789012345'
+        self.admin.assert_icommand(['iadmin', 'moduser', self.username, 'password', pw4], 'STDERR', 'PASSWORD_EXCEEDS_MAX_SIZE')
+
+        # Change password to 46 bytes (should fail)
+        pw4 = '1234567890123456789012345678901234567890123456'
+        self.admin.assert_icommand(['iadmin', 'moduser', self.username, 'password', pw4], 'STDERR', 'PASSWORD_EXCEEDS_MAX_SIZE')
```
Consider adding an assertion following this that proves the failed calls to modify the user's password don't result in breaking the user's ability to execute commands. The assertion can do an `ils` or something else entirely.

JustinKyleJames replied:
I added some code to verify that a session can be created with the last "good" password and not with the "bad" passwords.

JustinKyleJames replied:
I added an "ils" command after creating the session for the last good password.

alanking commented:
For some reason the PR is still linked to the issue. We'll need to figure out how to make that go away.

Please resolve the open conversation if you feel that it is addressed or does not need to be addressed. Then I think we can go ahead and squash this to taste (leave off the #s for now).

trel commented:
I don't think we're going to be able to 'remove' the connection - if the owner/admin of the repo AND the creator of the issue neither have buttons... I think we just have to watch it when it autocloses, and then reopen/reclose manually.

korydraughn reviewed (commented):
Changes look good.

Has this PR been run through the test suite?

JustinKyleJames commented:
> Changes look good.
> 
> Has this PR been run through the test suite?

I honestly don't remember as it has been a while.  I'll run it just to make sure.

korydraughn commented:
Sounds good.

JustinKyleJames commented:
> Changes look good.
> 
> Has this PR been run through the test suite?

I ran through all unit and CI tests.  They passed.

Before doing that I rebased onto 4-2-stable.  I also squashed and forced push.

JustinKyleJames commented:
Pounded.

JustinKyleJames commented:
pounded

korydraughn reviewed (approved):
Excellent.

d-w-moore commented:
I found that this commit breaks a [PRC test](https://github.com/irods/python-irodsclient/blob/57dbfc17aeb6627278106fc31bc61ed9ef680244/irods/test/user_group_test.py#L110) where the test relies on  CAT_PASSWORD_ENCODING_ERROR to be thrown on "alice's" (eve's?) attempt to detect the use of an invalid old-password value to change their password. On tip of 4-2-stable, we now get `PASSWORD_EXCEEDS_MAX_SIZE` thrown instead, even though the old pw value is much less than the maximum length ie. `(MAX_PASSWORD_LEN - 8)`

If the old-password value is merely incorrect, I think we should be flagging it as an incorrect/invalid user parameter or such, and not as "too long".

JustinKyleJames commented:
> merely incorrect, I think we should be flagging it as an incorrect/invalid user parameter or such, rather than being "too long"( which in this case, it certainly wasn't).

This is not entirely surprising.  I stated on the original issue that I can't guarantee that that encoding error was caused by the password being too long.  I wasn't sure because of the password obfuscating logic being in the middle.

d-w-moore commented:
If it helps, I do know that `__eq__` gets called from the most-derived class, so if  you have `class PathObj(str)`, then `"a"==PathObj("a")`  and `PathObj("a")=="a"` both evaluate to `True`, and both call `PathObj.__eq__` ... but you probably already knew that : )
